src/storage/src/hummock/compactor/fast_compactor_runner.rs

```diff
@@ -78,28 +80,51 @@ impl BlockStreamIterator {
         sstable: TableHolder,
         block_stream: BlockDataStream,
         task_progress: Arc<TaskProgress>,
+        existing_table_ids: HashSet<StateTableId>,
     ) -> Self {
         Self {
             block_stream,
-            next_block_index: 0,
             sstable,
             iter: None,
             task_progress,
+            existing_table_ids,
         }
     }
 
     /// Wrapper function for `self.block_stream.next()` which allows us to measure the time needed.
     async fn download_next_block(&mut self) -> HummockResult<Option<(Bytes, Vec<u8>, BlockMeta)>> {
-        let (data, _) = match self.block_stream.next_block_impl().await? {
-            None => return Ok(None),
-            Some(ret) => ret,
+        let (data, meta) = {
+            let block_data;
+            let mut block_meta;
+            loop {
+                // try next block
+                match self.block_stream.next_block_impl().await? {
+                    None => return Ok(None),
+                    Some((data, _uncompressed_size)) => {
+                        block_meta = self
+                            .block_stream
+                            .block_meta(self.block_stream.block_index())
```

Comment:
Are you sure this correct? Based on the BlockDataStream implementation, `block_index` is pointing to the next block not the returning block of `next_block_impl`

Reply:
I think the logic can be easily messed up by using tbe wrong block idx. Please add UTs to make sure things are working as expected.